File: C/quicksort_centinela.c
// Programa para buscar un elemento con centinela y ordenar el arreglo mediante quicksort

#include <stdio.h>

// Funcion para llenar un arreglo

void fill_array(int arr[], int size)
{
    register int i;
    int element;

    for (i = 0; i < size; i++)
    {
        printf("\nIngrese el valor del elemento %d ", i);
        scanf("%d", &element);
        arr[i] = element;
    }
    printf("\n");
}

// Funcion para imprimir un arreglo

void print_array(int arr[], int size)
{
    register int i;

    printf("\nLos elementos del arreglo son: [ ");
    
    for (i = 0; i < size; i++)
    {
        printf("%d ",arr[i]);
    }
    
    printf("]\n");
}

// Funcion para la busqueda por centinela

int sentinel_search(int arr[], int size, int v)
{
    register int i;

    arr[size] = v;

    for (i = 0; arr[i] != v; i++);
    if (i < size)
        return i;
    else
        return -1;
}

// Funcion para la particion del arreglo

int partition(int arr[], int start, int end)
{
    register int j;
    int i, aux, pivote;

    i = start - 1;
    pivote = arr[end];

    for (j = start; j <= end - 1; j++)
    {
        if (arr[j] < pivote)
        {
            i++;
            aux = arr[i];
            arr[i] = arr[j];
            arr[j] = aux;
        }
    }
    i++;
    aux = arr[i];
    arr[i] = arr[end];
    arr[end] = aux;

    return i;
}

// Funcion para el ordenamiento por quicksort

void quick_sort(int arr[], int start, int end)
{
    int pivote;

    if (start < end)
    {
        pivote = partition(arr, start, end);
        quick_sort(arr, start, pivote - 1);
        quick_sort(arr, pivote + 1, end);
    }
}

void main()
{
    int n, pos, value;

    printf("\nIngrese la cantidad de elementos del arreglo: ");
    scanf("%d", &n);

    int array[n];

    // Llamada a funcion para llenar el arreglo

    fill_array(array, n);
    print_array(array, n);

    // Llamada a funcion para buscar un elemento por centinela

    printf("\nIngrese el valor del elemento a buscar: \n");
    scanf("%d", &value);
    pos = sentinel_search(array, n, value);
    printf("\nLa posicion del elemento %d es: %d", value, pos);

    // Llamada a funcion para ordenar el arreglo mediante quicksort

    printf("\nOrdenando por quicksort...\n");
    quick_sort(array, 0, n - 1);
    print_array(array, n);
}

File: Python/quicksort_centinela.py
# Programa para buscar un elemento por centinela y ordenar la lista por quicksort

# Funcion para rellenar una lista

def fill_list(lista):
    length = int(input("\nIngrese la longitud de la lista: "))

    for i in range(0, length):
        element = int(input(f"\nIngresa el valor del elemento {i}: "))
        lista.append(element)

# Funcion para la busqueda por centinela

def sentinel_search(lista, size, v):

    lista.append(v)
    i = 0

    while lista[i] != v:
        i += 1

    if i < size:
        return i
    else:
        return -1

# Funcion para particionar la lista

def partition(lista, start, end):

    pivote = lista[end]
    i = start

    for j in range(start, end):
        if lista[j] < pivote:
            aux = lista[i]
            lista[i] = lista[j]
            lista[j] = aux
            i += 1

    aux = lista[i]
    lista[i] = lista[end]
    lista[end] = aux

    return i

# Funcion para el quicksort

def quick_sort(lista, start, end):

    if start < end:
        pivote = partition(lista, start, end)
        quick_sort(lista, start, pivote - 1)
        quick_sort(lista, pivote + 1, end)



# Programa main()

int_list = []

# Llamada a funcion para rellenar la lista

fill_list(int_list)
print(f"\nLa lista es: {int_list}")
list_size = len(int_list)

# Llamada a funcion para la busqueda por centinela

value = int(input("Ingrese el valor del elemento a buscar: "))
pos = sentinel_search(int_list, list_size, value)
print(f"\nLa posicion del elemento {value} es: {pos}")

# Removiendo el centinela de la lista

int_list.pop()

# Llamada a funcion para ordenar por quicksort

print("\nOrdenando por quicksort...")
quick_sort(int_list, 0, list_size - 1)
print(f"\nLa lista ordenada es: {int_list}")